src/transport/Dat.ts

```diff
@@ -1,92 +1,254 @@
-                logs.push("Done importing")
-                logs.push("Archive size: " + dat.archive.content.byteLength)
-                return callback(dat.key.toString('hex'), logs)
+    public createDatNode(id:string, archive:string|{}, options:{}): Promise<DatNode> {
+        let me = this
+        return new Promise<DatNode>(function(resolve, reject) {
+            if (me._datnodes[id]) {
+                return reject(new Error("DatNode with ID " + id + " already exists"))
+            } else {
+                me._datnodes[id] = new DatNode(id, archive, options)
+                return resolve(me._datnodes[id])
+            }
+        })
+    }
+
+    public disposeDatNode(id:string): Promise<DatNode> {
+        let me = this
+        return new Promise<DatNode>(function(resolve, reject) {
+            let datnode:DatNode = me.getDatNode(id);
+            delete me._datnodes[id];
+            if (datnode) {
+                try {
+                    datnode.close()
+                } catch (error) {
+                    console.error("Error while closing '" + id)
+                    // console.error(error)
+                    return reject(datnode)
+                }
+            }
+            return resolve(datnode)
+        })
+    }
+
+    public shutdown(): Promise<any> {
+        let me = this
+        var subPromises = []
+        Object.keys(me._datnodes).forEach(function (id) {
+            subPromises.push(me.disposeDatNode(id)
+                .catch(function(error) {
+                    console.error("Unexpected error while closing datnode " + id )
+                    // console.error(error)
+                }))
+        })
+        return Promise.all(subPromises)
+            .then(function() { return new Promise<void>(function (resolve, reject) {
+                if (Object.keys(me._datnodes).length > 0) {
+                    console.error("Not all dat nodes were closed during shutdown:")
+                    console.error(Object.keys(me._datnodes))
+                }
+                me._datnodes = {}
+                resolve()
+            })})
+            .catch(function(error) {
+                console.error("Unexpected error while shutting down datmanager")
+                console.error(error)
             })
-            progress.on('put', function (_src, dest) {
-                logs.push("Added: " + dest.name)
+    }
+}
+
+export class DatNode extends EventEmitter {
+    _id: string
+    _archive: string|{}
+    _options: {[index:string] : any} = {}
+    _dat: any
+    _stats:any
+
+    constructor(id:string, archive:string|{}, options:{[index:string] : any}) {
+        super()
+        this._id = id
+        this._archive = archive
+        this._options = options
+    }
+
+    public getArchiveKey() : string {
+        return this._dat.key.toString('hex')
+    }
+
+    public getID() {
+        return this._id
+    }
+
+    peerFound(): boolean {
+        return !!+this._dat.network.connected
```

Comment:
Question: What is the `!!+` for?

Reply:
This code was stolen from test in the Dat project. a '+' before a variable casts a truthy value to a numeric type. I'm not sure it's needed beyond the double bang; it may be overkill. I interpret it to mean the 'connected', 'connected', etc propoerties in the Dat lib are supposed to be numeric (i.e., '0' or '1') and this is a guard to cast it, even if it's not been set.

Reply:
Interesting, I may look into this in the near future if I have some free time, seems intriguing to me.